server/src/main/java/io/druid/server/emitter/EmitterMonitorProvider.java

```diff
@@ -19,18 +19,24 @@
 
 package io.druid.server.emitter;
 
-import com.fasterxml.jackson.annotation.JsonProperty;
-import org.joda.time.Period;
+import com.metamx.metrics.Monitor;
+import io.druid.guice.LazySingleton;
 
-/**
- */
-public class HttpEmitterConfig extends com.metamx.emitter.core.HttpEmitterConfig
+import javax.annotation.Nullable;
+
+@LazySingleton
+public class EmitterMonitorProvider
```

Comment:
Is this needed? Can a Monitor simply be added to `druid.monitoring.monitors`  instead?

Reply:
so basically my suggestion is adding in some monitors that can be used in `druid.monitoring.monitors` for the emitters in question, rather than having a special monitor case in the workflow.

Reply:
See https://github.com/druid-io/druid/pull/4973#discussion_r145853318. Also I think that more complicated code is better than requirement for extra config. I think that minimizing the amount of config and "good by default" behaviour is the top priority.

Reply:
I don't think most users would be able to act on these metrics in any meaningful way. Unless there's some kind of action people can take, then it is a set of metrics mostly used for debugging purposes or monitoring by advanced setups. Plus there's a lot of monitors that are very handy for standard use that we don't include by default.

There's also the consideration of code maintainability. Having special cases for different kinds of monitors in https://github.com/druid-io/druid/pull/4973/files#diff-cba74f4106709a96941c9a365e22eddcR104 is against the spirit of what that class is trying to do. The monitoring approach here also has a reverse-injection kind of approach where the thing which is used to inject is injected so the injection can back propagate to where it is used, which is a bit of a tangle for maintainability.

A cleaner impl would have a monitor which checks the injected emitter to see if it can appropriately monitor it, and if not just ignore any monitoring (maybe with a warning).

Reply:
> @drcrallen there will be a problem, that such emitter will be able to support only emitters, included into the druid codebase, namely HttpPostEmitter and ParametrizedUriEmitter, but not emitters, defined in private extensions. Do you think it's ok?

Are you asking if it's ok if a builtin emitter-monitor only works on builtin emitters? Just making sure I understand the question.

I think that's fine, as long as extensions defining their own emitters could also define their own emitter-monitors. (I think this should be possible already, even before this patch, but maybe I'm missing something)